### PR TITLE
keep the order of the typesense result when hydrating with doctrine

### PR DESCRIPTION
**Pull request by etienneleba**
doctrine order by id, with sql FIELD we can order the doctrine entities like typesense results

### CODE REVIEW

File: src/Finder/CollectionFinder.php [hunk 1]
```diff
@@ -37,7 +38,10 @@ private function hydrate($results)
         foreach ($results->getResults() as $result) {
             $ids[] = $result['document'][$primaryKeyInfos['documentAttribute']];
         }
-        $hydratedResults = $this->em->getRepository($this->collectionConfig['entity'])->findBy([$primaryKeyInfos['entityAttribute'] => $ids]);
+        $rsm = new ResultSetMappingBuilder($this->em);
+        $rsm->addRootEntityFromClassMetadata($this->collectionConfig['entity'], 'e');
+        $query = $this->em->createNativeQuery('SELECT * FROM client WHERE id IN ('.implode(', ', $ids).') ORDER BY FIELD(id,'.implode(', ', $ids).')', $rsm);
```
Comment (npotier) [inline src/Finder/CollectionFinder.php:43]: I guess `client` should be replaced by a variable, no ?

**Comment by etienneleba** [comment]
Oh yeah! You are right, I did my pull request too fast sorry..

File: src/Finder/CollectionFinder.php [hunk 2]
```diff
@@ -40,7 +40,8 @@ private function hydrate($results)
         }
         $rsm = new ResultSetMappingBuilder($this->em);
         $rsm->addRootEntityFromClassMetadata($this->collectionConfig['entity'], 'e');
-        $query = $this->em->createNativeQuery('SELECT * FROM client WHERE id IN ('.implode(', ', $ids).') ORDER BY FIELD(id,'.implode(', ', $ids).')', $rsm);
+        $tableName = $this->em->getClassMetadata($this->collectionConfig['entity'])->getTableName();
+        $query = $this->em->createNativeQuery('SELECT * FROM '.$tableName.' WHERE id IN ('.implode(', ', $ids).') ORDER BY FIELD(id,'.implode(', ', $ids).')', $rsm);
```
Comment (etienneleba) [inline src/Finder/CollectionFinder.php:44]: Do you always use ids to retrieve the entities or it's possible to use another field ? I'm not sure

Comment (npotier) [reply src/Finder/CollectionFinder.php:44]: For now, it's always the ids :)

Comment (etienneleba) [reply src/Finder/CollectionFinder.php:44]: Ok, so it should work, just a little issue, we have to handle when the array of ids is empty

Comment (npotier) [reply src/Finder/CollectionFinder.php:44]: Hello, I made a mistake in my previous answer : its not always the id that is used to query, it the `$primaryKeyInfos['entityAttribute']` field

Comment (etienneleba) [reply src/Finder/CollectionFinder.php:44]: Ok, so there is another issue, in the typesense bundle config I created this field in one of my collection : 
```
id: # Entity attribute name
          name: entity_id # Typesense attribute name
          type: primary # Attribute type
```
like this I can use it as default_sorting_field

but  when it's time to remove or update, the code is looking for a field id, so it was failing. In typesense the field "id" is generated automatically by typesense server because we didn't give it one. 

In typesense the default_sorting_field have to be an int32 or float and the id have to be a string. 

So I kept my entity_id field like this, and during the transform of doctrine entity to typesense document I'm adding a string field id which is the doctrine id. So, It's working, but I think we have to find a better way to do it.. Mostly if we want to handle primary key which are not called "id".

I hope I was clear :sweat_smile:

**Comment by etienneleba** [comment]
Ok so with this last commit the bundle work with this kind of config :
```
books:
      entity: 'App\Entity\Book'
      fields:
        id:
          name: id
          type: string
      default_sorting_field: entity_id
```

which is not really user-friendly but at least it's working, could you accept this pr for the moment, like this I can use composer. And I think we should open an issue to see how handle ids.

**Comment by etienneleba** [comment]
The typesense id field was generated by typesense cause we didn't give it one, by luck maybe it was working (if doctrine ids are equals to typesense generated ids), but now, I'm creating a field id for typesense, and I give it the value of the doctrine primary key casted to string.

**Comment by npotier** [comment]
Hello @etienneleba, thank you for your suggestions.

I've been thinking about what issues you where facing, and made some changes in order to handle the situation you encountered.

**TL;DR** you can use the following configuration in order to have your entity id used both as a Typesense id and a sortable field:

```yaml
    # config/packages/acseo_typesense.yml
    collections:
        books:                                     # Typesense collection name
            entity: 'App\Entity\Book'              # Doctrine Entity class
            fields: 
                # Keeping Database and Typesense synchronized with ids
                id:                                # Entity attribute name
                    name: id                       # Typesense attribute name
                    type: primary                  # Attribute type
                # Using again id as a sortable field (int32 required)
                sortable_id:
                    entity_attribute: id           # Entity attribute name forced
                    name: entity_id
                    type: int32
                ...            
            default_sorting_field: sortable_id    # Default sorting field. Must be int32 or float

```

This is a BC change, because the `primary` definition is now casted to a Typesense **string** type instead of an **int32**.

You cans use the dev-master or the v0.2 tag to use this feature.

**Comment by etienneleba** [comment]
Thx, I will test it 😊
